### PR TITLE
data: Allow Terraform module bundling via vfsgen

### PR DESCRIPTION
**Pull request by wking**
The new `data` package works two ways:

1. For devs, it loads the Terraform modules (now under `data/data`) from the disk.  If you execute your installer from a directory other than `data` (i.e. most of the time), you should set `OPENSHIFT_INSTALL_DATA` to point at the `data/data` directory.  That makes it easy to adjust the static files during development, because there's no need to rebuild after a tweak.

2. For users, we compile the Terraform modules into the executable.  That makes it easy to distribute the file (no separate pieces or wrapping tarball).

This is very similar to #329.  The differences are:

1. #329 always extracts from the tarball.  There's no "load from the disk at runtime" workflow.
2. I'm vendoring a different package ;).
3. The compiled-in binaries are slightly different sizes.  As of my b740463d688:

    ```console
    $ go generate ./data
    writing assets_vfsdata.go
    $ CGO_ENABLED=0 go build -tags release ./cmd/openshift-install
    $ ls -l openshift-install 
    -rwxr-xr-x. 1 trking trking 37933716 Sep 25 23:59 openshift-install
    ```

    Compared with #329's:

    ```console
    $ ./hack/build.sh
    $ ls -l bin/openshift-install 
    -rwxr-xr-x. 1 trking trking 39625166 Sep 26 00:00 bin/openshift-install
    ```

    But <2 MB isn't a large percentage change, so this is probably a wash.
4. I've broken the old installer by shifting its Terraform modules around, deleting symlinks, etc.

Dev builds look like:

```console
$ CGO_ENABLED=0 go build ./cmd/openshift-install
```

with invocations like:

```console
$ OPENSHIFT_INSTALL_DATA=data/data OPENSHIFT_INSTALL_CLUSTER_NAME=wking-testing ... ./openshift-install --dir wking --log-level debug cluster
```

Release builds look like:

```console
$ go generate ./data
writing assets_vfsdata.go
$ CGO_ENABLED=0 go build -tags release ./cmd/openshift-install
```

and the invocation is the same except you don't need to bother with `OPENSHIFT_INSTALL_DATA`.

CC @crawford

### CODE REVIEW

**Comment by wking** [comment]
> 4. I've broken the old installer by shifting its Terraform modules around, deleting symlinks, etc.

I've pushed 6af70ec7bab0b21aa69a495d97ae6d6b4a0df4c9 as an ugly hack to work around this until something like openshift/release#1677 lands.

**Comment by wking** [comment]
[e2e-aws][1]:

```
Waiting for router to be created ...
NAME STATUS ROLES AGE VERSION
ip-10-0-12-62.ec2.internal Ready master 4m
v1.11.0+d4cacc0 ip-10-0-17-43.ec2.internal Ready master 4m
v1.11.0+d4cacc0 ip-10-0-2-231.ec2.internal Ready bootstrap 4m
v1.11.0+d4cacc0 ip-10-0-34-168.ec2.internal Ready master 4m
v1.11.0+d4cacc0
NAME DESIRED CURRENT UP-TO-DATE AVAILABLE AGE
router 1 1 1 0 5s
Installation failed error: timed out waiting for the condition
2018/09/26 09:56:59 Container setup in pod e2e-aws failed,
```

But smoke passed, so

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/332/pull-ci-openshift-installer-master-e2e-aws/107/build-log.txt

**Comment by abhinavdahiya** [comment]
@wking 
> For devs, it loads the Terraform modules (now under data/data) from the disk. If you execute your installer from a directory other than data (i.e. most of the time), you should set OPENSHIFT_INSTALL_DATA

So this is requires always? :/

File: data/data/config.tf [hunk 1]
```diff
@@ -0,0 +1,336 @@
+terraform {
```
Comment (abhinavdahiya) [inline data/data/config.tf]: `data/data` smells :) can it be `data/terraform` ?

File: data/assets.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import (
+	"net/http"
+	"os"
+)
+
+// Assets contains project assets.
+var Assets http.FileSystem
+
+func init() {
+	dir := os.Getenv("OPENSHIFT_INSTALL_DATA")
```
Comment (abhinavdahiya) [inline data/assets.go]: can we use GOPATH to set reasonable default?

**Comment by wking** [comment]
> >  For devs, it loads the Terraform modules (now under data/data) from the disk. If you execute your installer from a directory other than data (i.e. most of the time), you should set OPENSHIFT_INSTALL_DATA
>
> So this is required always? :/

No, you could run from the `data` package ;).  But in general, yes.  If it bothers you, build with the compiled-in data?  I've updated `build.sh` to build that style by default.

File: data/unpack.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	}
+	defer file.Close()
+
+	info, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	if info.IsDir() {
+		os.Mkdir(base, 0777)
```
Comment (abhinavdahiya) [inline data/unpack.go]: does it have to be `0777`?

File: steps/assets/aws/config.tf [hunk 4]
```diff
@@ -1 +0,0 @@
-../../../config.tf
```
Comment (abhinavdahiya) [inline steps/assets/aws/config.tf]: Why can't we keep the symlink?

File: data/assets.go (hunk 2 above)
Comment (wking) [reply data/assets.go]: > can we use GOPATH to set reasonable default?

Maybe?  But that will break in Go 1.11 when we move to modules and toss out `GOPATH` ;).  Maybe export a default in your `~/.bashrc` or whatever?  Or compile it in?

File: data/unpack.go (hunk 3 above)
Comment (wking) [reply data/unpack.go]: > does it have to be 0777?

It doesn't *have* to be, but this is generic, and your `umask` will reduce it to whatever you feel is reasonable.

File: pkg/asset/cluster/cluster.go [hunk 5]
```diff
@@ -68,11 +65,24 @@ func (c *Cluster) Generate(parents map[asset.Asset]*asset.State) (*asset.State,
...
+	templateDir, err := terraform.FindStepTemplates(tmpDir, terraform.InfraStep, tfvars.Platform)
 	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errors.New("infra step not found; set OPENSHIFT_INSTALL_DATA to point to the data directory")
+		}
 		return nil, fmt.Errorf("error finding terraform templates: %v", err)
 	}
 
+	// take advantage of the new installer only having one step.
+	err = os.Rename(filepath.Join(tmpDir, "config.tf"), filepath.Join(templateDir, "config.tf"))
```
Comment (abhinavdahiya) [inline pkg/asset/cluster/cluster.go]: shouldn't this happen in `Unpack`?

File: steps/assets/aws/config.tf (hunk 4 above)
Comment (wking) [reply steps/assets/aws/config.tf]: > Why can't we keep the symlink?

We might be able to, but we don't need it for `openshift-install` (only one step :), and I wasn't sure how well `http.FileSystem` handled symlinks.

File: modules/aws/bootstrap/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
+# Bootstrap Module
```
Comment (abhinavdahiya) [inline modules/aws/bootstrap/README.md]: can you combine this with 5c536a24ce6860a32b9ba02eeef8d0a38ecc39ba ?

File: pkg/asset/cluster/cluster.go (hunk 5 above)
Comment (wking) [reply pkg/asset/cluster/cluster.go]: > shouldn't this happen in `Unpack`?

`Unpack` is generic, just dump the `http.FileSystem` to disk.  This is specific to the content that we happen to stick into `Unpack`.  And I was going to shuffle this around later, and not bother to write the step that we don't need (e.g. only unpack `modules` and `steps/infra/aws` for AWS).  But I hadn't worked that up yet.

File: modules/aws/bootstrap/README.md (hunk 6 above)
Comment (wking) [reply modules/aws/bootstrap/README.md]: > can you combine this with 5c536a2 ?

Yes.  I made 2ac5913 a separate commit so I could revert it once openshift/release#1677 lands.  But I can squash them if you like.  Thoughts?

**Comment by crawford** [comment]
Let's just get this thing in to remove the deadlock between this repo and openshift/release.

**Comment by wking** [comment]
Dunno why GitHub says golint is sad; [Prow says it's happy][1].

/test golint

[1]: https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_installer/332/pull-ci-openshift-installer-master-golint/112

**Pull request merged**